Add instructions for building custom blocks docs (#7979)
PrefectHQ/prefect: docs/collections/contribute.md

# Contribute

> 

Reviewer: I'd like to see some introduction text between headings, even if just a sentence. "You can contribute blocks and collections by following these steps." Or words to that effect is fine.

> ## Creating custom blocks

Reviewer: ```suggestion
## Create custom blocks
```

> ### Build a block

Reviewer: Suggest removing this heading since the difference between "create a block" and "build a block" is minimal.

Building your own custom block is simple!

1. Subclass from `Block`.
2. Add a description alongside an `Attributes` and `Example` section in the docstring.
3. Set a `_logo_url` to point to a relevant image.
4. Create the `pydantic.Field`s of the block with a type annotation, `default` or `default_factory`, and a short description about the field.

> 6. Define the methods of the block.

Reviewer: You can make these all 1. and the parser will number them correctly.

For example, this is how the [Secret block is implemented](https://github.com/PrefectHQ/prefect/blob/main/src/prefect/blocks/system.py#L76-L102):
```python
from pydantic import Field, SecretStr
from prefect.blocks.core import Block

class Secret(Block):
    """
    A block that represents a secret value. The value stored in this block will be obfuscated when
    this block is logged or shown in the UI.

    Attributes:
        value: A string value that should be kept secret.

    Example:
        ```python
        from prefect.blocks.system import Secret
        secret_block = Secret.load("BLOCK_NAME")

        # Access the stored secret
        secret_block.get()
        ```
    """

    _logo_url = "https://images.ctfassets.net/gm98wzqotmnx/5uUmyGBjRejYuGTWbTxz6E/3003e1829293718b3a5d2e909643a331/image8.png?h=250"

    value: SecretStr = Field(
        default=..., description="A string value that should be kept secret."
    )  # ... indicates it's a required field

    def get(self):
        return self.value.get_secret_value()
```

> ### Inheriting an existing block

Reviewer: We haven't been super consistent about this, but I'd like to move away from the gerund form and to an infinitive form when introducing a thing to do: Build instead of Building.

> ### Inheriting an existing block

Reviewer 2: I'd prefer if this section demonstrated how to use functionality from existing blocks via composition rather than inheritance. I think that block composition gives a better experience for the block developer since there are fewer caveats, and it's also better for the user since it enables easier reuse of configuration.

Author: Okay planning to leave it out for now since I don't have a good idea for composition.

class SecretDict(Secret):
    """
    A block that represents a secret dictionary. The dictionary values stored
    in this block will be obfuscated when this block is logged or shown in the UI.

    Attributes:
        value: A dictionary with its values that should be kept secret.

    Example:
        ```python
        from prefect.blocks.system import SecretDict
        secret_block = Secret.load("BLOCK_NAME")

        # Access the stored secret
        secret_block.get()
        ```
    """

    _block_type_name = "Secret Dictionary"
    _logo_url = "https://images.ctfassets.net/gm98wzqotmnx/5uUmyGBjRejYuGTWbTxz6E/3003e1829293718b3a5d2e909643a331/image8.png?h=250"

    value: SecretDict = Field(
        default=..., description="A dictionary with its values that should be kept secret."
    )  # ... indicates it's a required field
```
Just be sure to update `_block_type_name` and `_logo_url` as needed.

Then to use this:
```python
my_secret_dict = SecretDict(value={"a key": "to my deepest secrets"})
print(my_secret_dict)
print(my_secret_dict.get())
```

That should output:
```
SecretDict(value=SecretDict('{'a key': '**********'}'))
{'a key': 'to my deepest secrets'}
```

> To view on UI, [register the block](https://orion-docs.prefect.io/concepts/blocks/#registering-blocks-for-use-in-the-prefect-ui).

Reviewer: ```suggestion
To view in the Prefect Cloud or Prefect Orion server UI, [register the block](https://orion-docs.prefect.io/concepts/blocks/#registering-blocks-for-use-in-the-prefect-ui).
```

## Contributing Collections

Anyone can create and share a Prefect Collection and we encourage anyone interested in creating a collection to do so!

### Generating a project

To help you get started with your collection, we've created a template that gives the tools you need to create and publish your collection.

Use the [Prefect Collection template](https://github.com/PrefectHQ/prefect-collection-template#quickstart) to get started creating a collection with a bootstrapped project!

> ### Listing in the Collections Catalog

Reviewer: ```suggestion
### List a project in the Collections Catalog
```
